lib/puppet/provider/agent_upgrade/windows.rb

```diff
@@ -0,0 +1,85 @@
+
+Puppet::Type.type(:agent_upgrade).provide(:agent_upgrade) do
+  confine :osfamily => /windows/i
+
+  mk_resource_methods
+
+  commands :powershell =>
+             if File.exists?("#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe")
+               "#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe"
+             elsif File.exists?("#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe")
+               "#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe"
+             else
+               'powershell.exe'
+             end
+
+  def self.instances
+    [new({
+           :name => 'puppet-agent',
+           :ensure => :present,
+           :version => Facter.value(:puppetversion),
+         })
+    ]
+  end
+
+  def create
+    spawn_upgrade_process
+  end
+
+  def exists?
+    false
+  end
+
+  def source_url
+    resource[:source] ||
+      "https://downloads.puppetlabs.com/windows/#{puppet_agent_msi_filename}"
+  end
+
+  def puppet_agent_msi_filename
+    arch = resource[:arch] || Facter.value(:architecture)
+    if resource[:version] =~ /latest/i
+      return "puppet-agent-#{arch}-latest.msi"
+    end
+
+    "puppet-agent-#{resource[:version]}-#{arch}.msi"
+  end
+
+
+  def get_command
+    tempfile = new_logfile_path
+    debug "Spawning installer with log at '#{tempfile}'"
+    "cmd.exe /c \"\"#{native_path(command(:powershell))}\"" +
+      " -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command " +
+      "\"Wait-Process #{Process.pid} -ErrorAction SilentlyContinue;msiexec /qn /i #{source_url} /l*v #{tempfile} \""
+  end
+
+  def new_logfile_path
+    native_path(Tempfile.new(['puppet-agent-installer', '.log']).path)
+  end
+
+  def spawn_upgrade_process
+    args = {
+      :command_line => get_command,
+      :creation_flags => Process::DETACHED_PROCESS,
+      :process_inherit => false,
+      :thread_inherit => true,
+      :cwd => 'C:\\'
+    }
+
+    debug "Creating wait process for #{Process.pid}"
+    upgradeProcess = Process.create(args)
+    debug ("Created background process to wait for puppet exit at PID #{upgradeProcess.process_id}")
```

Comment:
"We are going to upgrade the Puppet Agent as soon as Puppet finishes running the catalog. Please be patient and give it about 10-15 minutes to complete. We will log the progress in *logfilelocation* and the detailed installation log will be at *msiloglocation*. If the node doesn't come back up, please check one or both of these log files to determine what may have went wrong."

As far as what support would also want to state in there, best to ask them.